### PR TITLE
gpg: add initial bare-bones wrapper

### PR DESCRIPTION
**Pull request by DomT4**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

++++
Split out of https://github.com/Homebrew/brew/pull/386 because trying to bundle everything together in one huge PR to submit like a block of 🍰 was turning up some pretty significant problems & simultaneously making me unhappy and making others think I've been sat here smoking a cigar this whole time.
++++

This is the first diddy steps of the wrapper, not the whole thing, obviously.

I don't want to dump the whole functionality from other branches yet because A) as mentioned above, merging unique branches together has revealed some not great problems with some of the existing logic not included here, B) this is the absolute minimum required for the core PR to be in a good place to ship, and people are getting anxious with nothing having been shipped yet, including me.

### CODE REVIEW

File: Library/Homebrew/gpg.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+require "utils"
+
+class Gpg
+  # Should ideally be using `GPGRequirement.new.gpg2`, etc to get path here but
+  # calling that directly leads to:
+  # requirement.rb:139:in `which_all': uninitialized constant Requirement::ORIGINAL_PATHS (NameError)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/gpg.rb]: You could/should probably do the reverse and have the `GPGRequirement` use the code from here. If that still doesn't work, shout and I'll try to get it working.

File: Library/Homebrew/gpg.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      gpg2_short_version = Utils.popen_read(gpg2, "--version")[/\d\.\d/, 0]
+      next unless gpg2_short_version
+      Version.create(gpg2_short_version.to_s) == Version.create("2.0")
+    end
+  end
+
+  GPG_EXECUTABLE = gpg2 || gpg
+
+  def self.available?
+    File.exist?(GPG_EXECUTABLE.to_s) && File.executable?(GPG_EXECUTABLE)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/gpg.rb]: I think `File.executable?` may check existence without barfing. May also want to check it's not a directory.

**Comment by MikeMcQuaid** [comment]
A few comments but generally 👍

File: Library/Homebrew/gpg.rb (hunk 1 above)
Comment (DomT4) [inline Library/Homebrew/gpg.rb]: I thought about that. Wasn't sure if it was acceptable or not, but if it is, sounds reasonable.

I had some issues around `which_all` always returning a `Pathname` even if you forced `to_s` on it but I think I've just worked out why that wasn't working. Will double check with a fuller test locally.

File: Library/Homebrew/gpg.rb (hunk 2 above)
Comment (DomT4) [inline Library/Homebrew/gpg.rb]: `which_all` (which I've switched back to locally) already does [some checks](https://github.com/Homebrew/brew/blob/604f4b6a0785e0295e548e6d033823303da19061/Library/Homebrew/utils.rb#L430) the file is executable & a file. This is mostly a smoke test so we actually have something for `available?`.

**Comment by DomT4** [comment]
Pushed changes. Obviously it just gets skipped on CI, but the test does pass:

```
GpgTest#test_create_test_key = 0.31 s = .

Finished in 81.197964s, 11.3919 runs/s, 30.0008 assertions/s.

  1) Skipped:
IntegrationCommandTests#test_bundle [/usr/local/Library/Homebrew/test/test_integration_cmds.rb:42]:
HOMEBREW_TEST_OFFICIAL_CMD_TAPS is not set


  2) Skipped:
IntegrationCommandTests#test_services [/usr/local/Library/Homebrew/test/test_integration_cmds.rb:42]:
HOMEBREW_TEST_OFFICIAL_CMD_TAPS is not set


  3) Skipped:
IntegrationCommandTests#test_pull [/usr/local/Library/Homebrew/test/test_integration_cmds.rb:765]:
Requires network connection


  4) Skipped:
IntegrationCommandTests#test_cask [/usr/local/Library/Homebrew/test/test_integration_cmds.rb:42]:
HOMEBREW_TEST_OFFICIAL_CMD_TAPS is not set

925 runs, 2436 assertions, 0 failures, 0 errors, 4 skips
```

**Comment by MikeMcQuaid** [comment]
:shipit: 👍

**Pull request closed without merging**